File: python/divergent-modifications/good/report_generator.py
class ReportGenerator:
    def __init__(self, transaction_repository, calculator):
        self.transaction_repository = transaction_repository
        self.calculator = calculator

    def generate_monthly_report(self, user_id: int, month: int, year: int):
        return self.transaction_repository.get_monthly_transactions(user_id, month, year)

    def generate_tax_report(self, user_id: int, year: int):
        income = self.transaction_repository.get_yearly_income(user_id, year)
        deductions = self.transaction_repository.get_yearly_deductions(user_id, year)
        tax_owed = self.calculator.calculate_tax(income, deductions)

        return {
            'year': year,
            'total_income': income,
            'total_deductions': deductions,
            'tax_owed': tax_owed
        }

File: golang/divergent-modifications/good/report_generator.go
package main

type reportGenerator struct {
	transactionRepository TransactionRepository
	calculator           FinancialCalculator
}

func NewReportGenerator(transactionRepo TransactionRepository, calculator FinancialCalculator) ReportGenerator {
	return &reportGenerator{
		transactionRepository: transactionRepo,
		calculator:           calculator,
	}
}

func (rg reportGenerator) GenerateMonthlyReport(userId, month, year int) []map[string]interface{} {
	return rg.transactionRepository.GetMonthlyTransactions(userId, month, year)
}

func (rg reportGenerator) GenerateTaxReport(userId, year int) map[string]interface{} {
	income := rg.transactionRepository.GetYearlyIncome(userId, year)
	deductions := rg.transactionRepository.GetYearlyDeductions(userId, year)
	taxOwed := rg.calculator.CalculateTax(income, deductions)

	return map[string]interface{}{
		"year":             year,
		"total_income":     income,
		"total_deductions": deductions,
		"tax_owed":         taxOwed,
	}
}
